<a href="https://colab.research.google.com/github/kellytog/BI/blob/main/Reto_Challenge_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Primero cargo las librerias de trabajo

In [ ]:
import numpy as np
import pandas as pd

1. cargo la base de datos

In [ ]:
nxl ='/content/1. BD1_Longitudinal Survey_Int.xlsx'
XDB=pd.read_excel(nxl)
XDB.head(10)

#selecciono variables de interes
XDB=XDB[['Age', 'Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize','Urban']]
XDB.head(10)


#separo variables de entrada y salida
XD=XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']]
YD=XDB['Urban']

XD = XD.dropna()
YD = YD.dropna()

XD = XD[XD.index.isin(YD.index)]
YD = YD[YD.index.isin(XD.index)]


In [ ]:
# Número total de datos en la base
print("Número total de datos en la base:", XDB.shape[0])

ct = mnb.class_count_
print("Como están separados los datos", ct)
print("El total de datos es:", len(XD))
print("Los datos porcentualmente están divididos:", ct/len(XD))

Número total de datos en la base: 12686
Como están separados los datos [1239. 4709.]
El total de datos es: 5948
Los datos porcentualmente están divididos: [0.20830531 0.79169469]


In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB()
mnb.fit(XD,YD)

#obtenemos la info de edad
u=mnb.theta_
print("la media de la edad es:\n",u)
var=mnb.var_
sigma=np.sqrt(var)
print("la desviacion estandar de la edad es:\n",sigma)

#determino lim sup e inf de edad
LS=u+sigma
LI=u-sigma
print("el limite superior de la edad es:\n",LS)
print("el limite inferior de la edad es:\n",LI)


la media de la edad es:
 [[1.83640032e+01 1.07312349e+01 1.02711864e+01 4.73930589e+00
  2.19951574e+01 6.72429379e+01 1.49003228e+02 3.01248854e+04]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 4.68103631e+00
  2.25744319e+01 6.69116585e+01 1.44750903e+02 3.32736742e+04]]
la desviacion estandar de la edad es:
 [[2.47260545e+00 3.15661127e+00 3.88837565e+00 2.55226748e+00
  4.21650807e+00 4.08228636e+00 3.03486350e+01 3.16482374e+04]
 [2.47344359e+00 3.40066236e+00 4.11549881e+00 2.44205851e+00
  4.19763081e+00 4.20193102e+00 3.00285301e+01 3.60897302e+04]]
el limite superior de la edad es:
 [[2.08366087e+01 1.38878461e+01 1.41595621e+01 7.29157338e+00
  2.62116655e+01 7.13252242e+01 1.79351863e+02 6.17731228e+04]
 [2.08353039e+01 1.45620554e+01 1.53486693e+01 7.12309482e+00
  2.67720628e+01 7.11135895e+01 1.74779433e+02 6.93634044e+04]]
el limite inferior de la edad es:
 [[ 1.58913978e+01  7.57462359e+00  6.38281079e+00  2.18703841e+00
   1.77786493e+01  6.31606515e+01  1.18654593e+

In [ ]:
#obtenemos la info de altura
u=mnb.theta_
print("la media de la altura es:\n",u)
var=mnb.var_
sigma=np.sqrt(var)
print("la desviacion estandar de la altura es:\n",sigma)

#determino lim sup e inf de altura
LS=u+sigma
LI=u-sigma
print("el limite superior de la altura es:\n",LS)
print("el limite inferior de la altura es:\n",LI)


la media de la altura es:
 [[1.83640032e+01 1.07312349e+01 1.02711864e+01 4.73930589e+00
  2.19951574e+01 6.72429379e+01 1.49003228e+02 3.01248854e+04]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 4.68103631e+00
  2.25744319e+01 6.69116585e+01 1.44750903e+02 3.32736742e+04]]
la desviacion estandar de la altura es:
 [[2.47260545e+00 3.15661127e+00 3.88837565e+00 2.55226748e+00
  4.21650807e+00 4.08228636e+00 3.03486350e+01 3.16482374e+04]
 [2.47344359e+00 3.40066236e+00 4.11549881e+00 2.44205851e+00
  4.19763081e+00 4.20193102e+00 3.00285301e+01 3.60897302e+04]]
el limite superior de la altura es:
 [[2.08366087e+01 1.38878461e+01 1.41595621e+01 7.29157338e+00
  2.62116655e+01 7.13252242e+01 1.79351863e+02 6.17731228e+04]
 [2.08353039e+01 1.45620554e+01 1.53486693e+01 7.12309482e+00
  2.67720628e+01 7.11135895e+01 1.74779433e+02 6.93634044e+04]]
el limite inferior de la altura es:
 [[ 1.58913978e+01  7.57462359e+00  6.38281079e+00  2.18703841e+00
   1.77786493e+01  6.31606515e+01  1.18

In [ ]:
#obtenemos la info del peso de las personas
u=mnb.theta_
print("la media del peso es:\n",u)
var=mnb.var_
sigma=np.sqrt(var)
print("la desviacion estandar del peso es:\n",sigma)

#determino lim sup e inf de peso
LS=u+sigma
LI=u-sigma
print("el limite superior del peso es:\n",LS)
print("el limite inferior del peso es:\n",LI)

#evaluamos un solicitante
XDS=[[34,12,15,2,24,69,160,48000]]
ydp=mnb.predict(XDS) # se obtiene la prediccion
print("la prediccion del solicitante es:\n",ydp)

if ydp==0:
  print("el solicitante no es urbano")
if ydp==1:
  print("el solicitante es urbano")


la media del peso es:
 [[1.83640032e+01 1.07312349e+01 1.02711864e+01 4.73930589e+00
  2.19951574e+01 6.72429379e+01 1.49003228e+02 3.01248854e+04]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 4.68103631e+00
  2.25744319e+01 6.69116585e+01 1.44750903e+02 3.32736742e+04]]
la desviacion estandar del peso es:
 [[2.47260545e+00 3.15661127e+00 3.88837565e+00 2.55226748e+00
  4.21650807e+00 4.08228636e+00 3.03486350e+01 3.16482374e+04]
 [2.47344359e+00 3.40066236e+00 4.11549881e+00 2.44205851e+00
  4.19763081e+00 4.20193102e+00 3.00285301e+01 3.60897302e+04]]
el limite superior del peso es:
 [[2.08366087e+01 1.38878461e+01 1.41595621e+01 7.29157338e+00
  2.62116655e+01 7.13252242e+01 1.79351863e+02 6.17731228e+04]
 [2.08353039e+01 1.45620554e+01 1.53486693e+01 7.12309482e+00
  2.67720628e+01 7.11135895e+01 1.74779433e+02 6.93634044e+04]]
el limite inferior del peso es:
 [[ 1.58913978e+01  7.57462359e+00  6.38281079e+00  2.18703841e+00
   1.77786493e+01  6.31606515e+01  1.18654593e+02 -1.52

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


In [ ]:
from sklearn.metrics import confusion_matrix

#haga un barrido de todos los datos aquellos que pueden quedar mal
ydp2=mnb.predict(XD)  #hagael pronostico para los datos base

cm=confusion_matrix(YD,ydp2) #cree la matriz de confusion
print("la matriz de confusion es:\n",cm)
VN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]

#metricas desempeño
Ex=(VP+VN)/(VP+VN+FP+FN) #EXACTITUD: comportamiento general preaprobando
print("la exactitud del modelo es:\n",Ex)

TE=(FN+FP)/(VP+VN+FP+FN) #TASA DE ERROR: comportamiento general mal preaprobando
print("la tasa de error del modelo es:\n",TE)

Se=VP/(VP+FN) #SENSIBILIDAD: comportamiento correcto preaprobando sobre el total de los preaporbados
print("la sensibilidad del modelo es:\n",Se)

Sp=VN/(VN+FP) #ESPECIFICIDAD: comportamiento correcto preaprobando sobre el total de los no preaporbados
print("la especificidad del modelo es:\n",Sp)

Pr=VP/(VP+FP) #PRECISION: comportamiento correcto preaprobando sobre el total de los preaporbados
print("la precision del modelo es:\n",Pr)

PrNeg=VN/(VN+FN) #PRECISION NEGATIVA: comportamiento correcto preaprobando sobre el total de los no preaporbados
print("la precision negativa del modelo es:\n",PrNeg)

la matriz de confusion es:
 [[   3 1236]
 [   8 4701]]
la exactitud del modelo es:
 0.7908540685944856
la tasa de error del modelo es:
 0.20914593140551446
la sensibilidad del modelo es:
 0.9983011255043533
la especificidad del modelo es:
 0.002421307506053269
la precision del modelo es:
 0.7918140474987367
la precision negativa del modelo es:
 0.2727272727272727


De acuerdo con la base de datos suministrada, tomamos como variables 'Age', 'Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize','Urban'. La base de datos posee un total de 5948 datos en total divididos en 21% rurales y 79% urbanos.
El modelo alcanzo una exactitud del 79% respecto a su tasa de error del 20%; ademas es importante resaltar su precision de 79% ya que nos muestra que el modelo tiene muy buen comportamiento frente a la toma de decisiones